server/user.js

```diff
@@ -0,0 +1,330 @@
+user.updateSession = function(cookie) {
+  var db = database.db;
+  var query = {$and: [{username: cookie.username},
+    {sessionId: cookie.sessionId}]};
+  if (cookie.sessionId) {
+    // Have session and update expire date.
+    var newExpiration = new Date().getTime() + user.cookieExpireTime;
+    db.collection(user.sessionDbName).update(cookie,
+      {$set: {expiration: newExpiration}}, {upsert: false});
+    cookie.expiration = newExpiration;
+  } else {
+    // Don't have username or have username but don't have valid session
+    cookie.sessionId = utils.randomString();
+    cookie.expiration = new Date().getTime() + user.cookieExpireTime;
+    db.collection(user.sessionDbName).insertOne(cookie);
+  }
+  return cookie;
+};
+
+/**
+ * Authenticates the signed up user.
+ * @param {Array<!user.Info>} data
+ * @param {!user.Info} userInfo User Information.
+ * @return {user.Error|undefined}
+ * @private
+ */
+user.signUpUser_ = function(data, userInfo) {
+  var db = database.db;
+  var duplicates = [];
+  data.forEach(function(item) {
+    if (item.email == userInfo.email) {
+      duplicates.push('email: ' + userInfo.email);
+    }
+    if (item.username == userInfo.username) {
+      duplicates.push('username: ' + userInfo.username);
+    }
+  });
+  if (duplicates.length) {
+    var errorMessage = duplicates.join(' and ') + ' exist(s)';
+    return {
+      error: errorMessage
+    };
+  } else {
+    db.collection(user.userInfoDbName).insertOne(userInfo);
+    log.serverLog('user information saved');
+  }
+};
+
+/**
+ * Validates email address.
+ * @param {string} email
+ * @return {boolean}
+ * @private
+ */
+user.validateEmail_ = function(email) {
+  return utils.validateRegex(email, user.VALID_EMAIL_REGEX);
+};
+
+/**
+ * Validates username, allows letters, numbers, and underscores, and no less
+ * than 6 characters.
+ * @param {string} username
+ * @return {boolean}
+ * @private
+ */
+user.validateUsername_ = function(username) {
+  return utils.validateRegex(username, user.VALID_USERNAME_REGEX);
+};
+
+/**
+ * Validates password, allows letters, numbers, and underscores, and no less
+ * than 8 characters.
+ * @param {string} password
+ * @return {boolean}
+ * @private
+ */
+user.validatePassword_ = function(password) {
+  return utils.validateRegex(password, user.VALID_PASSWORD_REGEX);
+};
+
+/**
+ * Authenticates the signed up user.
+ * @param {string} username Username.
+ * @param {string} password Password.
+ * @param {string=} email Email.
+ * @return {boolean}
+ */
+user.validateUserInfo = function(username, password, email) {
+  var invalidItem = [];
+  if (!user.validateUsername_(username)) {
+    invalidItem.push(username);
+  }
+  if (!user.validatePassword_(password)) {
+    invalidItem.push(password);
+  }
+  if (email && !user.validateEmail_(email)) {
+    invalidItem.push(email);
+  }
+  if (invalidItem.length) {
+    var errorMessage = invalidItem.join(' and ') + ' is(are) invalid';
+    log.serverLog(errorMessage);
```

Comment:
This seems to be a utility function. It could be better to consistently log error in callers. You can provide surrounding environments in the log, e.g.
```js
log.serverLog('cannot login user', errorMessage);
log.serverLog('cannot auto-login user', errorMessage);
```